### PR TITLE
[SC359] Add queries for reporting category overview

### PR DESCRIPTION
**Pull request by seerocode**
### **Changes:**
1. Added reporting queries for getting the following category overview for an optional date range
overview

<details>
  <summary>Example for category ID and a date range:</summary>
  
```grapqhl
query GetCategoryOverview($id: ID!, $dateRange: DateRange) {
    categoryOverview(id: $id, dateRange: $dateRange) {
    id 
    name
    description
    sumCategoryValues{
      categoryValueId
      categoryValue
      sum
    }
  }
}

```

```
{"id": "category_id_here", 
  "dateRange": 
  {"start_date": "2021-01-31 00:00:00", 
  "end_date": "2023-12-31 00:00:00"}
}
```
Result

```
{
  "data": {
    "categoryOverview": {
      "id": "51349e29-290e-4398-a401-5bf7d04af75e",
      "name": "Gender",
      "description": "Gender: A social construct based on a group of emotional and psychological characteristics that classify an individual as feminine, masculine, androgynous or other. Gender can be understood to have several components, including gender identity, gender expression and gender role.",
      "sumCategoryValues": [
        {
          "categoryValueId": "1525cce8-7db3-4e73-b5b0-d2bd14777534",
          "categoryValue": "Trans men",
          "sum": 2
        },
        {
          "categoryValueId": "662557e5-aca8-4cec-ad72-119ad9cda81b",
          "categoryValue": "Trans women",
          "sum": 2
        },
        {
          "categoryValueId": "6cae6d26-97e1-4e9c-b1ad-954b4110e83b",
          "categoryValue": "Non-binary",
          "sum": 20
        },
        {
          "categoryValueId": "742b5971-eeb6-4f7a-8275-6111f2342bb4",
          "categoryValue": "Cisgender women",
          "sum": 8
        },
        {
          "categoryValueId": "a72ced2b-b1a6-4d3d-b003-e35e980960df",
          "categoryValue": "Gender non-conforming",
          "sum": 2
        },
        {
          "categoryValueId": "d237a422-5858-459c-bd01-a0abdc077e5b",
          "categoryValue": "Cisgender men",
          "sum": 8
        }
      ]
    }
  }
}
```
</details>

<details>
  <summary>Example for all categories and date range:</summary>

```
query GetCategoryOverviews($dateRange: DateRange) {
    categoriesOverview(dateRange: $dateRange) {
    id 
    name
    description
    sumCategoryValues {
      categoryValueId
      categoryValue
      sum
    }
  }
}
```

```
{"dateRange": 
  {"start_date": "2021-01-31 00:00:00", 
  "end_date": "2023-12-31 00:00:00"}
}
```
</details>

2. Add reporting query for header counts:
overview

<details>
<summary>Example query:</summary>

```
query GetReportingOverview($dateRange: DateRange) {
  reportingOverview(dateRange: $dateRange) {
    countData {
      teamsCount
      datasetsCount
      tagsCount
    }
    categoriesOverview {
      name
    }
  }
}

```

Result:
```
{
  "data": {
    "reportingOverview": {
      "countData": {
        "teamsCount": 2,
        "datasetsCount": 2,
        "tagsCount": 1
      },
      "categoriesOverview": [
        {
          "name": "Gender"
        },
        {
          "name": "Race / ethnicity"
        },
        {
          "name": "Disability"
        }
      ]
    }
  }
}
```
</details>

### CODE REVIEW

**Comment by seerocode** [comment]
These analytics queries might make more sense as graphql subscriptions with table views that are cached when possible. I'll leave as queries for now.

**Comment by seerocode** [comment]
@nmanduj1 I'm refactoring a little bit here to make the queries more performant.

File: api/database.py [hunk 1]
```diff
@@ -214,6 +213,10 @@ class Team(Base, PermissionsMixin):
                      server_default=func.now(), onupdate=func.now())
     deleted = Column(TIMESTAMP)
 
+    @classmethod
+    def get_unfiltered_count(cls, session):
```
Comment (nmanduj1) [inline api/database.py:217]: naming clarification to make sure I understand- what's an `unfiltered_count` vs a `count` ?

File: api/graphql/reporting.schema.graphql [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# overview
+type ReportingOverview @needsPermission(permission: [ADMIN, TEAM_MEMBER]) {
```
Comment (nmanduj1) [inline api/graphql/reporting.schema.graphql]: IIRC, I think we're only setting up the reporting dash to be viewable by lara-levels admins as the MVP

File: api/reporting_queries.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            CategoryValue.name.label('category_value_name'),
+            CategoryValue.id.label('category_value_id'),
+            func.sum(Entry.count).label('sum_of_counts')).\
+                join(CategoryValue.entries).\
+                join(Entry.record).\
+                    group_by(CategoryValue.id).\
+                    filter(CategoryValue.category_id == category.id, Entry.deleted == None, CategoryValue.deleted == None)
+
+    # optional start and end date params passed to categoryOverview
+    if 'dateRange' in info.variable_values:
```
Comment (nmanduj1) [inline api/reporting_queries.py:70]: Nice !

File: api/graphql/reporting.schema.graphql (hunk 2 above)
Comment (seerocode) [reply api/graphql/reporting.schema.graphql]: Thanks for catching this!

**Comment by nmanduj1** [comment]
This looks good to me- tho I'm def no whiz at setting up analytic queries.  I trust your judgement and think we'd be good to merge and iterate as necessary. Great job!  🚢🚢

**Review by nmanduj1 — APPROVED** [review]

File: api/database.py (hunk 1 above)
Comment (seerocode) [reply api/database.py:217]: Unfiltered would mean we get a count that includes every single team no matter if it's part of our parent filtered search or not.

**Comment by seerocode** [comment]
@nmanduj1 Ok to review. I'll continue any performance checks later. Not important now!

**Comment by seerocode** [comment]
> This looks good to me- tho I'm def no whiz at setting up analytic queries. I trust your judgement and think we'd be good to merge and iterate as necessary. Great job! 🚢🚢

🥳  There's def room for refactoring but I'd like to get this out first and update as we go.

**Pull request merged**